### PR TITLE
Make sure the correct block template file is used in the Site Editor for templates with fallback

### LINKED ISSUE #48489: Product Archive templates (Products by * and Product Search Results) load Product Catalog by default
**Issue Description:**
On fresh sites the blockified templates of:
- Product by Category ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/taxonomy-product_cat.html))
- Product by Tag ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/taxonomy-product_tag.html))
- Product by Attribute ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/taxonomy-product_attribute.html))
- Product Search Results ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/product-search-results.html))
are not loaded and instead the fallback Product Catalog ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/archive-product.html)) is loaded directly.

Products by * are coincidentally the same as Product Catalog so the issue wasn't spotted but there are more significant differences for Product Search Results so the problem is more visible.

**Steps to reproduce**
0. You can introduce some changes to blockified templates to make the changes visible:
  - Product by Category ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/taxonomy-product_cat.html))
  - Product by Tag ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/taxonomy-product_tag.html))
  - Product by Attribute ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/taxonomy-product_attribute.html))
  - Product Search Results ([file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/templates/blockified/product-search-results.html))
 
e.g.
```
	<!-- wp:heading -->
	<h2 class="wp-block-heading"> Template Name </h2>
	<!-- /wp:heading -->
```

1. Create a fresh site (the key part is the templates above were not customized)
2. Go to Editor and visit these templates
3. **Expected:** Changes you introduced are visible (correct template is loaded)
4. **Actual:** There's no changes and fallback Product Catalog template is visible

image

5. Edit the template (literally anything, you can add a paragraph with some text) and save
6. Click "Clear customizations"

image

7. Now you can see the correct template is loaded (the one that's expected in step 3.)

image

### PR DESCRIPTION
**Pull request by Aljullu**
### Changes proposed in this Pull Request:

Fixes #48489.

This PR fixes the logic that was handling the template file retrieval for templates which fall back to another template. As mentioned in #48489, we were always loading the fallback template file (aka, Product Catalog), instead of the specific template file. That was a problem for the Product Search Results template, as it has a different file than Product Catalog.

Sidenote: in https://github.com/woocommerce/woocommerce/issues/48536 we are investigating next steps to make sure the Product Search Results template doesn't fall back to the Product Catalog template.

### How to test the changes in this Pull Request:

1. From `example.com/wp-admin/`, go to Appearance > Site Editor > Templates and open the Product Search Results template. (Note: it's important to open the template by clicking on the links mentioned, instead of navigating directly to the template page using the URL).
2. Verify the correct template is loaded. You can notice it by the title that is used:

Before | After
--- | ---
imatge | imatge

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/tests/e2e/tests/templates/product-search-results.block_theme.spec.ts
Comment (Aljullu) [inline plugins/woocommerce-blocks/tests/e2e/tests/templates/product-search-results.block_theme.spec.ts:1]: As mentioned in https://github.com/woocommerce/woocommerce/pull/48436, the issue was happening for all templates that "fall back" to Product Catalog, but in practice, it was only visible in the Product Search Results template, as it was the only template with different contents than Product Catalog. That's why I only wrote the test for the Product Search Results template.

File: plugins/woocommerce-blocks/tests/e2e/tests/templates/product-search-results.block_theme.spec.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		editor,
+		page,
+	} ) => {
+		const templateName = 'Product Search Results';
+		await admin.visitSiteEditor();
+		await page.getByRole( 'button', { name: 'Templates' } ).click();
+		await page.getByPlaceholder( 'Search' ).fill( templateName );
+		// Wait until search has finished.
+		const searchResults = page.getByLabel( 'Actions' );
+		await expect.poll( async () => await searchResults.count() ).toBe( 1 );
```
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/templates/product-search-results.block_theme.spec.ts]: This is an equivalent to:
```suggestion
		await expect( searchResults ).toHaveCount( 1 );
```

Comment (Aljullu) [reply plugins/woocommerce-blocks/tests/e2e/tests/templates/product-search-results.block_theme.spec.ts]: Much nicer, thanks! I merged it. :slightly_smiling_face:

**Review by WunderBart — COMMENTED** [review]
I'm trying to repro OnMyMachine™, but I see the `after` version of the template/title on both `trunk` and this branch: 

Screenshot 2024-06-21 at 13 48 29

I'm testing against TT4, FWIW.

**Review by WunderBart — APPROVED** [review]
Alright, I can repro now - the trick is to actually perform the navigation to that template instead just refreshing the page between branches. Apparently, the template loads differently in both cases. Weird bug, and a nice catch! 😄 

🚀

**Pull request merged**